### PR TITLE
Add possibility to add bcs on edges/vertices

### PR DESCRIPTION
**Pull request by lijas**
Useful for for example "simply supported" bcs. 
Also useful for shell elements which only has edges.

See Cantilever.ipynb for example usage

### CODE REVIEW

File: src/Dofs/ConstraintHandler.jl [hunk 1]
```diff
@@ -145,7 +157,7 @@ function _add!(ch::ConstraintHandler, dbc::Dirichlet, bcfaces::Set{Tuple{Int,Int
     # loop over all the faces in the set and add the global dofs to `constrained_dofs`
     constrained_dofs = Int[]
     #_celldofs = fill(0, ndofs_per_cell(ch.dh))
-    for (cellidx, faceidx) in bcfaces
+    for  (cellidx, faceidx) in bcfaces
```
Comment (KristofferC) [inline src/Dofs/ConstraintHandler.jl:160]:
```suggestion
    for (cellidx, faceidx) in bcfaces
```

File: docs/src/literate/incompressible_elasticity.jl [hunk 2]
```diff
@@ -157,7 +157,7 @@ function assemble_up!(Ke, fe, cell, cellvalues_u, cellvalues_p, facevalues_u, gr
     ## We loop over all the faces in the cell, then check if the face
     ## is in our `"traction"` faceset.
     @inbounds for face in 1:nfaces(cell)
-        if onboundary(cell, face) && (cellid(cell), face) ∈ getfaceset(grid, "traction")
+        if onboundary(cell, face) && FaceIndex(cellid(cell), face) ∈ getfaceset(grid, "traction")
```
Comment (KristofferC) [inline docs/src/literate/incompressible_elasticity.jl:160]: Hm, I guess this is pretty breaking. Perhaps we can leave `FaceIndex ` as `const FaceIndex = Tuple{Int, Int}` but use struct for the others?

File: examples/cantilever.ipynb [hunk 3]
```diff
@@ -244,42 +212,18 @@
    "cell_type": "code",
    "execution_count": null,
    "metadata": {},
-   "outputs": [
-    {
-     "name": "stdout",
-     "output_type": "stream",
-     "text": [
-      " \u001b[1m------------------------------------------------------------------\u001b[22m\n",
-      "                           Time                   Allocations      \n",
```
Comment (KristofferC) [inline examples/cantilever.ipynb:253]: Did this table go away?

File: src/Dofs/ConstraintHandler.jl [hunk 4]
```diff
@@ -229,6 +241,7 @@ function _update!(values::Vector{Float64}, f::Function, faces::Set{Tuple{Int,Int
     _celldofs = fill(0, ndofs_per_cell(dh, _tmp_cellid))
 
     for (cellidx, faceidx) in faces
+
```
Comment (KristofferC) [inline src/Dofs/ConstraintHandler.jl:244]:
```suggestion
```

File: src/Dofs/ConstraintHandler.jl [hunk 5]
```diff
@@ -395,22 +409,30 @@ end
 
 #Function for adding constraint when using multiple celltypes
 function add!(ch::ConstraintHandler, fh::FieldHandler, dbc::Dirichlet)
-    dbc.faces isa Set{Tuple{Int}} && error("Can only apply boundary conditions to faces (not nodes).")
+
```
Comment (KristofferC) [inline src/Dofs/ConstraintHandler.jl:412]:
```suggestion
```

**Comment by lijas** [comment]
I added a simple example with shell elements, which is the strongest motivation why one needs to be able to add BCs on edges/vertices. I am planning to write some text for the example so that it can be added to the other tutorials. 

I would  also like some feedback on the structure of the code. Currently in the ` add!`-function for the `ConstraintHandler`, I changed it so that it creates `BCValues` for either faces, edges or vertices:

```
function add!(ch::ConstraintHandler, dbc::Dirichlet)
...
functype = getgeometryfunction(eltype(dbc.bcset)) # eltype(dbc.bcset) is either FaceIndex, EdgeIndex or VertexIndes
bcvalue = BCValues(interpolation, JuAFEM.default_interpolation(celltype), functype)
_add!(ch, dbc, dbc.bcset, interpolation, field_dim, field_offset(ch.dh, dbc.field_name), bcvalue, functype)
```

where

```
getgeometryfunction(::Type{FaceIndex}) = JuAFEM.faces
getgeometryfunction(::Type{EdgeIndex}) = JuAFEM.edges
getgeometryfunction(::Type{VertexIndex}) = JuAFEM.vertices
```

Do you think it is clear what  `getgeometryfunction` is doing? Is there a better name? Or should we skip it all together and just do

```
if eltype(dbc.bcset) isa FaceIndex
    functype = JuAFEM.faces
elseif eltype(dbc.bcset) isa EdgeIndex
   functype = JuAFEM.edges
elseif eltype(dbc.bcset) isa VertexIndex
   functype = JuAFEM.vertices
end
bcvalue = BCValues(interpolation, JuAFEM.default_interpolation(celltype), functype)
_add!(ch, dbc, dbc.faces, interpolation, field_dim, field_offset(ch.dh, dbc.field_name), bcvalue, functype)
```

Any feedback would be appriciated..

**Comment by KristofferC** [comment]
I think doing the explicit `if` makes sense for now. If the same type of code occurs in other places, it could be changed out to dispatch.

**Comment by lijas** [comment]
Close due to #343

**Pull request closed without merging**